wrapper/src/main/java/software/amazon/jdbc/ConnectionInvalidator.java

```diff
@@ -0,0 +1,167 @@
+            return invalidateThread;
+          });
+  private static final ExecutorService abortConnectionExecutorService =
+      Executors.newCachedThreadPool(
+          r -> {
+            final Thread abortThread = new Thread(r);
+            abortThread.setDaemon(true);
+            return abortThread;
+          });
+
+  private static final Logger LOGGER = Logger.getLogger(ConnectionInvalidator.class.getName());
+  private static final RdsUtils rdsUtils = new RdsUtils();
+
+  public static void populateOpenedConnectionQueue(final HostSpec hostSpec, final Connection conn) {
+    final Set<String> aliases = hostSpec.getAliases().isEmpty()
+        ? new HashSet<>(Collections.singletonList(hostSpec.asAlias()))
+        : hostSpec.getAliases();
+
+    // Check if the connection was established using a cluster endpoint
+    final RdsUrlType urlType = rdsUtils.identifyRdsType(hostSpec.getHost());
+    String endpoint = null;
+    if (urlType.isRdsCluster()) {
+      LOGGER.info("Attempting to populate opened connection queue with cluster endpoint.");
+
+      // 1. Check if aliases contain an instance endpoint.
+      LOGGER.info(String.format("host instance: %s\n", hostSpec.getHost()));
+
+      for (String alias : aliases) {
+        LOGGER.info(String.format("alias: %s\n", alias));
+        if (!rdsUtils.identifyRdsType(alias).isRdsCluster()) {
+          endpoint = alias;
+          break;
+        }
+      }
+
+      if (endpoint == null) {
+        LOGGER.info("Unable to populate opened connection queue with cluster endpoints.");
+        return;
+      }
+    }
+
+    Queue<WeakReference<Connection>> connectionQueue = getConnectionQueue(aliases);
+    connectionQueue.add(new WeakReference<>(conn));
+
+    for (String alias : aliases) {
+      if (!rdsUtils.identifyRdsType(alias).isRdsCluster()) {
+        openedConnections.put(alias, connectionQueue);
+      }
+    }
+
+    logOpenedConnections();
+  }
+
+  /**
+   * Invalidates all opened connections pointing to the same node in a daemon thread. 1
+   *
+   * @param hostSpec The {@link HostSpec} object containing the url of the node.
+   */
+  public static void invalidateAllConnections(final HostSpec hostSpec) {
+    invalidateAllConnections(hostSpec.getAliases().toArray(new String[] {}));
+  }
+
+  public static void invalidateAllConnections(final String... node) {
+    Queue<WeakReference<Connection>> connectionQueue = getConnectionQueue(
+        new HashSet<>(Arrays.asList(node)));
+
+    if (connectionQueue.isEmpty()) {
+      return;
+    }
+
+    invalidateConnections(connectionQueue);
+  }
+
+  private static Queue<WeakReference<Connection>> getConnectionQueue(Set<String> aliases) {
+    Queue<WeakReference<Connection>> connectionQueue;
+
+    for (String alias : aliases) {
+      if (openedConnections.containsKey(alias)) {
+        connectionQueue = openedConnections.get(alias);
+        if (connectionQueue != null) {
+          return connectionQueue;
```

Comment:
Should we continue with loop? Other aliases may give other queues of connections.